src/promnesia/sources/reddit.py

```diff
@@ -3,40 +3,153 @@
+        except ImportError as import_err:
+            # TODO: add dummy _link_extractor and _parser_cls classes incase
+            # these are called by a subclass?
+
+            # only send error if the user is trying to enable this feature
+            if render_markdown:
+                logger.exception(import_err)
+                logger.critical("Could not import markdown module to render reddit markdown. Try 'python3 -m pip install mistletoe'")
+            render_markdown = False  # force to be false, couldn't import
+        self.render_markdown = render_markdown
+
+
+    def _from_comment(self, i: 'Comment') -> Results:
+        locator = Loc.make(
+            title='Reddit comment',
+            href=i.url,
+        )
+        yield from self._from_common(i, locator=locator)
+
+
+    def _from_submission(self, i: 'Submission') -> Results:
+        locator = Loc.make(
+            title=f'Reddit submission: {i.title}',
+            href=i.url,
+        )
+        yield from self._from_common(i, locator=locator)
+
+
+    def _from_upvote(self, i: 'Upvote') -> Results:
+        locator = Loc.make(
+            title=f'Reddit upvote',
+            href=i.url,
+        )
+        yield from self._from_common(i, locator=locator)
+
+
+    def _from_save(self, i: 'Save') -> Results:
+        locator = Loc.make(
+            title='Reddit save',
+            href=i.url,
+        )
+        yield from self._from_common(i, locator=locator)
+
+
+    # to allow for possible subclassing by the user?
+    def _render_body(self, text: str) -> str:
+        if self.render_markdown:
+            return self._parser_cls(text)._doc_ashtml()
+        else:
+            return text
+
+
+    def _from_common(self, i: 'RedditThing', locator: Loc) -> Results:
+        urls = [i.url]
+        # TODO this should belong to HPI.. fix permalink handling I guess
+        # ok, it's not present for all of them..
+        lurl = i.raw.get('link_url')
+        if lurl is not None:
+            urls.append(lurl)
+        lurl = i.raw.get('url')
+        if lurl is not None:
+            urls.append(lurl)
+
+        context = self._render_body(i.text)
+
+        emitted: Set[str] = set()
+
+        for url in chain(urls, extract_urls(i.text)):
+            if url in emitted:
+                continue
+            yield Visit(
+                url=url,
+                dt=i.created,
+                context=context,
+                locator=locator,
+            )
+            emitted.add(url)
+
+        # extract from markdown links like [link text](https://...)
+        # incase URLExtract missed any
+        #
+        # this should try to do this, even if the user didn't enable
+        # the render_markdown flag, as it may catch extra links that URLExtract didnt
+        # would still require mistletoe to be installed, but
+        # the user may already have it installed for the auto/markdown modules
+        if hasattr(self, '_link_extractor'):
```

Comment:
maybe instead of hasattr, make _link_extractor optional and None if it's not in use?